File: kube/src/main/java/com/jacksonrakena/infrastructure/envs/prod/ProductionStack.kt
package com.jacksonrakena.infrastructure.envs.prod

import com.jacksonrakena.infrastructure.apps.GradekeeperServer
import com.jacksonrakena.infrastructure.apps.Identity
import com.jacksonrakena.infrastructure.apps.Jacksonbot
import com.jacksonrakena.infrastructure.apps.Mixer
import com.jacksonrakena.infrastructure.apps.Mxbudget
import com.jacksonrakena.infrastructure.apps.persistence.Bouncer
import com.jacksonrakena.infrastructure.apps.persistence.Galahad
import com.jacksonrakena.infrastructure.traefik.TraefikStack
import com.jacksonrakena.infrastructure.util.loadTlsSecretFromFolder
import org.cdk8s.ApiObjectMetadata
import org.cdk8s.Chart
import org.cdk8s.ChartProps
import org.cdk8s.plus28.HttpIngressPathType
import org.cdk8s.plus28.Ingress
import org.cdk8s.plus28.IngressBackend
import org.cdk8s.plus28.IngressProps
import org.cdk8s.plus28.IngressRule
import org.cdk8s.plus28.IngressTls
import org.cdk8s.plus28.TlsSecret
import software.constructs.Construct

class ProductionStack(
    scope: Construct,
    id: String,
    props: ChartProps,
) : Chart(scope, id, props) {
    val credentials = ProductionCredentials(this, "credentials", props)

    val storage = ProductionBlockStorage(this, id, props)

    val galahad = Galahad(
        this,
        "galahad",
        credentials.vaultwardenSecret,
        credentials.postgresSecret,
        storage.volumeClaim,
        props
    )

    val bouncer = Bouncer(this, "bouncer", galahad, props)

    val tf = TraefikStack(this, "traefik", props)
    val gks = GradekeeperServer(
        this,
        "gk-server",
        credentials.gradekeeperConfigMap,
        galahad.postgresService,
        credentials.githubRegistrySecret
    )

    val ident = Identity(this, "id", credentials.keycloakConfigMap, galahad, props)

    val jacksonbot =
        Jacksonbot(
            this,
            "jacksonbot",
            credentials.jacksonbotConfigMap,
            credentials.githubRegistrySecret,
            galahad.postgresService,
            props
        )

    val mx = Mixer(this, "mixer", credentials.mixerConfigMap, credentials.githubRegistrySecret, props)
    val mxbudget = Mxbudget(
        this,
        "mxbudget",
        credentials.githubRegistrySecret,
        bouncer.service,
        credentials.mxbudgetConfigMap,
        props
    )

    val rakenaComAuTlsSecret =
        TlsSecret(
            this, "rakena-com-au-cert",
            loadTlsSecretFromFolder("secrets/cert-rakena.com.au")
        )
    val rakenaCoNzTldSecret = TlsSecret(
        this, "rakena-co-nz-cert",
        loadTlsSecretFromFolder("secrets/cert-rakena.co.nz")
    )

    val ingress = Ingress(
        this, "ingress", IngressProps.builder()
            .metadata(
                ApiObjectMetadata.builder()
                    .annotations(
                        mapOf(
                            "traefik.ingress.kubernetes.io/router.tls" to "true",
                            "traefik.ingress.kubernetes.io/router.entrypoints" to "websecure"
                        )
                    ).build()
            )
            .tls(
                listOf(
                    IngressTls.builder()
                        .secret(rakenaComAuTlsSecret)
                        .hosts(
                            listOf(
                                "id.rakena.com.au",
                                "budget.rakena.com.au",
                                "vault.rakena.com.au"
                            )
                        )
                        .build(),
                    IngressTls.builder()
                        .secret(rakenaCoNzTldSecret)
                        .hosts(
                            listOf(
                                "vault.rakena.co.nz"
                            )
                        ).build()
                )
            )
            .rules(
                listOf(
                    IngressRule.builder()
                        .host("budget.rakena.com.au")
                        .pathType(HttpIngressPathType.PREFIX)
                        .backend(IngressBackend.fromService(mxbudget.service))
                        .build(),
                    IngressRule.builder()
                        .host("id.rakena.com.au")
                        .pathType(HttpIngressPathType.PREFIX)
                        .backend(IngressBackend.fromService(ident.service))
                        .build(),
                    IngressRule.builder()
                        .host("vault.rakena.com.au")
                        .pathType(HttpIngressPathType.PREFIX)
                        .backend(IngressBackend.fromService(galahad.vaultService))
                        .build(),
                    IngressRule.builder()
                        .host("vault.rakena.co.nz")
                        .pathType(HttpIngressPathType.PREFIX)
                        .backend(IngressBackend.fromService(galahad.vaultService))
                        .build(),
                    IngressRule.builder()
                        .host("api.gradekeeper.xyz")
                        .pathType(HttpIngressPathType.PREFIX)
                        .backend(IngressBackend.fromService(gks.service))
                        .build(),
                )
            )
            .build()
    )
}

File: src/envs/prod/production-stack.ts
import { Construct } from "constructs";
import { Chart, ChartProps } from "cdk8s";
import * as kplus from "cdk8s-plus-28";
import { ProductionCredentials } from "./production-credentials";
import { ProductionBlockStorage } from "./production-block-storage";
import { Vault } from "../../apps/vault";
import { TraefikStack } from "../../traefik/traefik-stack";
import { GradekeeperServer } from "../../apps/gradekeeper-server";
import { Blank } from "../../apps/blank";
import { loadTlsSecretFromFolder } from "../../util/secret-utils";
import { Mixer } from "../../apps/mixer";
import { LeodeCluster } from "../../apps/persistence/leode";
import {
  Gateway,
  GatewaySpecListenersAllowedRoutesNamespacesFrom,
  GatewaySpecListenersTlsMode,
  HttpRoute,
  HttpRouteSpecRulesMatchesPathType,
} from "../../../imports/gateway.networking.k8s.io";
import { createPostgresRoleAndSecret } from "../../util/pg-utils";

export class ProductionStack extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps) {
    super(scope, id, props);

    const credentials = new ProductionCredentials(this, "credentials", props);

    const storage = new ProductionBlockStorage(this, id, props);

    const vault = new Vault(
      this,
      "vault",
      credentials.vaultwardenSecret,
      storage.volumeClaim,
      props,
    );

    const gkRole = createPostgresRoleAndSecret(this, "gradekeeper");
    const mixerRole = createPostgresRoleAndSecret(this, "mixer");
    const leode = new LeodeCluster(
      this,
      "leode",
      storage.ociFreeStorageClass,
      [gkRole, mixerRole].map((e) => e.role),
      props,
    );

    new TraefikStack(this, "traefik", props);

    const gks = new GradekeeperServer(
      this,
      "gk-server",
      credentials.gradekeeperConfigMap,
      leode.services.readWrite,
      gkRole.secret,
      credentials.githubRegistrySecret,
      props,
    );

    const mixer = new Mixer(
      this,
      "mixer",
      credentials.mixerBackendConfigMap,
      credentials.githubRegistrySecret,
      mixerRole.secret,
      props,
    );

    const rakenaComAuTlsSecret = new kplus.TlsSecret(
      this,
      "rakena-com-au-cert",
      loadTlsSecretFromFolder("secrets/cert-rakena.com.au"),
    );
    const rakenaCoNzTlsSecret = new kplus.TlsSecret(
      this,
      "rakena-co-nz-cert",
      loadTlsSecretFromFolder("secrets/cert-rakena.co.nz"),
    );
    const jacksonrakenaComTlsSecret = new kplus.TlsSecret(
      this,
      "jacksonrakena-com-cert",
      loadTlsSecretFromFolder("secrets/cert-jacksonrakena.com"),
    );

    const blank = new Blank(
      this,
      "blank",
      credentials.githubRegistrySecret,
      props,
    );

    const gateway = new Gateway(this, "gateway", {
      metadata: {
        name: "traefik-gateway",
        namespace: this.namespace,
      },
      spec: {
        gatewayClassName: "traefik-gateway-class",
        infrastructure: {
          annotations: {
            "oci.oraclecloud.com/load-balancer-type": "nlb",
          },
        },
        listeners: [
          {
            name: "https-rakenaconz",
            protocol: "HTTPS",
            port: 443,
            hostname: "*.rakena.co.nz",
            tls: {
              mode: GatewaySpecListenersTlsMode.TERMINATE,
              certificateRefs: [rakenaCoNzTlsSecret].map((secret) => ({
                name: secret.name,
                namespace: this.namespace,
              })),
            },
            allowedRoutes: {
              namespaces: {
                from: GatewaySpecListenersAllowedRoutesNamespacesFrom.SAME,
              },
            },
          },
          {
            name: "https-jacksonrakenacom",
            protocol: "HTTPS",
            port: 443,
            hostname: "*.jacksonrakena.com",
            tls: {
              mode: GatewaySpecListenersTlsMode.TERMINATE,
              certificateRefs: [jacksonrakenaComTlsSecret].map((secret) => ({
                name: secret.name,
                namespace: this.namespace,
              })),
            },
            allowedRoutes: {
              namespaces: {
                from: GatewaySpecListenersAllowedRoutesNamespacesFrom.SAME,
              },
            },
          },
          {
            name: "https-rakenacomau",
            protocol: "HTTPS",
            port: 443,
            tls: {
              mode: GatewaySpecListenersTlsMode.TERMINATE,
              certificateRefs: [rakenaComAuTlsSecret].map((secret) => ({
                name: secret.name,
                namespace: this.namespace,
              })),
            },
            allowedRoutes: {
              namespaces: {
                from: GatewaySpecListenersAllowedRoutesNamespacesFrom.SAME,
              },
            },
          },
        ],
      },
    });

    const createRoute = (
      name: string,
      hostname: string,
      service: kplus.Service,
    ) =>
      new HttpRoute(this, name, {
        metadata: {
          name,
          namespace: this.namespace,
        },
        spec: {
          parentRefs: [
            {
              name: "traefik-gateway",
            },
          ],
          hostnames: [hostname],
          rules: [
            {
              matches: [
                {
                  path: {
                    type: HttpRouteSpecRulesMatchesPathType.PATH_PREFIX,
                    value: "/",
                  },
                },
              ],
              backendRefs: [
                {
                  name: service.name,
                  port: service.ports[0].port,
                },
              ],
            },
          ],
        },
      });

    [
      {
        name: "vault-route",
        hostname: "vault.rakena.com.au",
        service: vault.service,
      },
      {
        name: "vault-co-nz-route",
        hostname: "vault.rakena.co.nz",
        service: vault.service,
      },
      {
        name: "gradekeeper-route",
        hostname: "api.gradekeeper.xyz",
        service: gks.service,
      },
      {
        name: "jacksonrakena-route",
        hostname: "go.jacksonrakena.com",
        service: blank.service,
      },
      {
        name: "finance-api-route",
        hostname: "finance-api.rakena.com.au",
        service: mixer.service,
      },
      {
        name: "finance-frontend-route",
        hostname: "finance.rakena.com.au",
        service: mixer.frontendService,
      },
    ].map(({ name, hostname, service }) =>
      createRoute(name, hostname, service),
    );
  }
}
